File: android/testSrc/com/android/tools/idea/templates/GradleFileMergerTest.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.templates;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectManager;
import com.intellij.openapi.util.io.FileUtil;
import org.jetbrains.android.AndroidTestCase;
import org.jetbrains.annotations.Nullable;

import java.io.File;

/**
 * Test gradle merging by PSI tree implementation
 */
public class GradleFileMergerTest extends AndroidTestCase {

  public void testProjectDisposal() throws Exception {
    Project[] openProjects = ProjectManager.getInstance().getOpenProjects();
    checkFileMerge("templates/Base.gradle", "templates/NewFlavor.gradle", "templates/MergedNewFlavor.gradle", null);
    Project[] postMergeOpenProjects = ProjectManager.getInstance().getOpenProjects();
    assertFalse(postMergeOpenProjects.length > openProjects.length);
    for (Project p : postMergeOpenProjects) {
      if (p.getName().equals("MergingOnly")) {
        fail();
      }
    }
  }

  public void testInsertFlavor() throws Exception {
    checkFileMerge("templates/Base.gradle",
                   "templates/NewFlavor.gradle",
                   "templates/MergedNewFlavor.gradle");
  }

  public void testInsertBuildType() throws Exception {
    checkFileMerge("templates/Base.gradle",
                   "templates/NewBuildType.gradle",
                   "templates/MergedNewBuildType.gradle");
  }

  public void testMergeDependencies() throws Exception {
    checkDependencyMerge("templates/Base.gradle", "templates/NewDependencies.gradle", "templates/MergedNewDependencies.gradle");
  }

  public void testMergeCloudDependencies() throws Exception {
    checkDependencyMerge("templates/Base.gradle", "templates/CloudDependencies.gradle", "templates/MergedCloudDependencies.gradle");
  }

  public void testMergeCloudDependenciesDuplicate() throws Exception {
    checkDependencyMerge("templates/Base.gradle", "templates/CloudDependenciesDuplicate.gradle",
                         "templates/MergedCloudDependenciesDuplicate.gradle");
  }

  public void testMergeCloudDependenciesExclude() throws Exception {
    checkDependencyMerge("templates/BaseExclude.gradle", "templates/CloudDependencies.gradle",
                         "templates/MergedCloudDependenciesExclude.gradle");
  }

  private void checkDependencyMerge(String destPath, String srcPath, String goldenPath) {
    File destFile = new File(getTestDataPath(), FileUtil.toSystemDependentName(destPath));
    String dest = TemplateUtils.readTextFromDisk(destFile);
    assertNotNull(dest);

    File srcFile = new File(getTestDataPath(), FileUtil.toSystemDependentName(srcPath));
    String source = TemplateUtils.readTextFromDisk(srcFile);
    assertNotNull(source);

    File goldenFile = new File(getTestDataPath(), FileUtil.toSystemDependentName(goldenPath));
    String golden = TemplateUtils.readTextFromDisk(goldenFile);
    assertNotNull(golden);

    // Strip comments from merged file
    assertEquals(golden.replaceAll("\\s+","\n"),
                 GradleFileMerger.mergeGradleFiles(source, dest, getProject(), null).replaceAll("\\s+//.*", "").replaceAll("\\s+", "\n"));
  }

  public void testRemapFlavorAssetDir() throws Exception {
    checkFileMerge("templates/Base.gradle",
                   "templates/RemapFlavorAssetDir.gradle",
                   "templates/MergedRemapFlavorAssetDir.gradle");
  }

  public void testRenameAndroidManifestFile() throws Exception {
    checkFileMerge("templates/Base.gradle",
                   "templates/RenameManifest.gradle",
                   "templates/MergedRenameManifest.gradle");
  }

  private void checkFileMerge(@Nullable String destPath, @Nullable String srcPath, @Nullable String goldenPath) throws Exception {
    checkFileMerge(destPath, srcPath, goldenPath, getProject());
  }

  private static void checkFileMerge(@Nullable String destPath, @Nullable String srcPath, @Nullable String goldenPath,
                                     @Nullable Project project) throws Exception {
    String source = "";
    String dest = "";
    String golden = "";
    if (destPath != null) {
      File destFile = new File(getTestDataPath(), FileUtil.toSystemDependentName(destPath));
      dest = TemplateUtils.readTextFromDisk(destFile);
      assertNotNull(dest);
    }

    if (srcPath != null) {
      File srcFile = new File(getTestDataPath(), FileUtil.toSystemDependentName(srcPath));
      source = TemplateUtils.readTextFromDisk(srcFile);
      assertNotNull(source);
    }

    if (goldenPath != null) {
      File goldenFile = new File(getTestDataPath(), FileUtil.toSystemDependentName(goldenPath));
      golden = TemplateUtils.readTextFromDisk(goldenFile);
      assertNotNull(golden);
    }

    assertEquals(golden.replaceAll("\\s+","\n"),
                 GradleFileMerger.mergeGradleFiles(source, dest, project, null).replaceAll("\\s+", "\n"));
  }
}

File: android/testSrc/com/android/tools/idea/templates/GradleFileMergerTest.kt
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.templates

import com.android.ide.common.repository.GradleCoordinate
import com.android.tools.idea.templates.GradleFilePsiMerger.mergeGradleFiles
import com.android.tools.idea.testing.AndroidGradleTestCase
import com.google.common.base.Charsets
import com.google.common.collect.LinkedListMultimap
import com.google.common.collect.Multimap
import com.google.common.io.Files
import com.google.common.truth.Truth.assertThat
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.util.io.FileUtil
import java.io.File

class GradleFileMergerTest : AndroidGradleTestCase() {
  fun testProjectDisposal() {
    val openProjects = ProjectManager.getInstance().openProjects
    checkFileMerge("Base.gradle", "NewFlavor.gradle", "MergedNewFlavor.gradle")
    val postMergeOpenProjects = ProjectManager.getInstance().openProjects
    assertThat(openProjects.size).isAtLeast(postMergeOpenProjects.size)
    if (postMergeOpenProjects.any {it.name == "MergingOnly"}) {
      fail()
    }
  }

  fun testInsertFlavor() =
    checkFileMerge("Base.gradle", "NewFlavor.gradle", "MergedNewFlavor.gradle")

  fun testInsertBuildType() =
    checkFileMerge("Base.gradle", "NewBuildType.gradle", "MergedNewBuildType.gradle")

  fun testMergeDependencies() =
    checkFileMerge("Base.gradle", "NewDependencies.gradle", "MergedNewDependencies.gradle")

  fun testMergeCloudDependencies() =
    checkFileMerge("Base.gradle", "CloudDependencies.gradle", "MergedCloudDependencies.gradle")

  fun testMergeCloudDependenciesDuplicate() =
    checkFileMerge("Base.gradle", "CloudDependenciesDuplicate.gradle", "MergedCloudDependenciesDuplicate.gradle")

  fun testMergeCloudDependenciesExclude() =
    checkFileMerge("BaseExclude.gradle", "CloudDependencies.gradle", "MergedCloudDependenciesExclude.gradle")

  fun testDifferentConfigurationDependencies() {
    checkFileMerge("Base.gradle", "TestDependencies.gradle", "MergedTestDependencies.gradle")
    checkFileMerge( "MergedTestDependencies.gradle", "NewDependencies.gradle", "MergedTestAndNewDependencies.gradle" )
  }

  fun testRemapFlavorAssetDir() =
    checkFileMerge("Base.gradle", "RemapFlavorAssetDir.gradle", "MergedRemapFlavorAssetDir.gradle")

  fun testRenameAndroidManifestFile() =
    checkFileMerge("Base.gradle", "RenameManifest.gradle", "MergedRenameManifest.gradle")

  fun testBuildscriptMerge() =
    checkFileMerge("BaseToplevel.gradle", "ToplevelInject.gradle", "MergedToplevelInject.gradle")

  fun testBuildscriptMergeIncludingSingleValueAssignment() {
    // To test the gradle files merge that include a single value assignment with the equal identifier
    checkFileMerge(
      "BaseToplevelIncludingSingleValueAssignment.gradle",
      "ToplevelInject.gradle",
      "MergedToplevelIncludingSingleValueAssignment.gradle"
    )
  }

  fun testAddNewDependenciesBlock() =
    checkFileMerge("AddDependenciesBlock.gradle", "NewDependencies.gradle", "MergedDependenciesBlock.gradle")

  fun testAddApplyPlugin() =
    checkFileMerge("Base.gradle", "AddApplyPlugin.gradle", "MergedApplyPlugin.gradle")

  fun testRemoveExistingDependencies() {
    checkDependenciesRemoved("compile", "compile")
    checkDependenciesRemoved("compile", "implementation")
    checkDependenciesRemoved("compile", "api")
    checkDependenciesRemoved("compile", "feature")
    checkDependenciesRemoved("implementation", "compile")
    checkDependenciesRemoved("implementation", "implementation")
    checkDependenciesRemoved("implementation", "api")
    checkDependenciesRemoved("implementation", "feature")
    checkDependenciesRemoved("testCompile", "testImplementation")
    checkDependenciesRemoved("testCompile", "testApi")
    checkDependenciesRemoved("androidTestCompile", "androidTestImplementation")
    checkDependenciesRemoved("androidTestCompile", "androidTestApi")
  }

  private fun checkFileMerge(destPath: String?, srcPath: String?, goldenPath: String?) {
    var source = ""
    var dest = ""
    var golden = ""
    if (destPath != null) {
      val destFile = File(getTestDataPath(), FileUtil.toSystemDependentName("templates/$destPath"))
      dest = Files.asCharSource(destFile, Charsets.UTF_8).read()
    }

    if (srcPath != null) {
      val srcFile = File(getTestDataPath(), FileUtil.toSystemDependentName("templates/$srcPath"))
      source = Files.asCharSource(srcFile, Charsets.UTF_8).read()
    }

    if (goldenPath != null) {
      val goldenFile = File(getTestDataPath(), FileUtil.toSystemDependentName("templates/$goldenPath"))
      golden = Files.asCharSource(goldenFile, Charsets.UTF_8).read()
    }

    val result = mergeGradleFiles(source, dest, project, null)

    assertEquals(golden.replace("\\s+".toRegex(), "\n"), result.replace("\\s+".toRegex(), "\n"))
  }

  private fun checkDependenciesRemoved(dstConfigName: String, srcConfigName: String) {
    val dependencyId = "com.android.support:appcompat-v7"
    val dependencyCoordinate = GradleCoordinate.parseCoordinateString("$dependencyId:23.1.1")

    val dstConfigs = LinkedListMultimap.create<String, GradleCoordinate>()
    dstConfigs.put(dependencyId, dependencyCoordinate)
    val dstAllConfigs = hashMapOf<String, Multimap<String, GradleCoordinate>>()
    dstAllConfigs[dstConfigName] = dstConfigs

    val srcConfigs = LinkedListMultimap.create<String, GradleCoordinate>()
    srcConfigs.put(dependencyId, dependencyCoordinate)
    val srcAllConfigs = hashMapOf<String, Multimap<String, GradleCoordinate>>()
    srcAllConfigs[srcConfigName] = srcConfigs

    updateExistingDependencies(srcAllConfigs, dstAllConfigs, null, null)

    assertThat(dstConfigs).hasSize(1)
    assertThat(srcConfigs).hasSize(0)
  }
}
